src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_agent_info_management.c

```diff
@@ -143,20 +108,61 @@ void OnMessage_ainfo_resp(
         goto done;
     }
 
-    // NOTE: Do no parse ainfo_resp msg payload as it is empty.
+    // NOTE: Do no parse ainfo_resp msg payload as it is empty as per client/server protocol design.
 
-    if (!Handle_AgentInfo_Response(
-        agentInfoData,
-        context))
+    if (!Handle_AgentInfo_Response(agentInfoData, retriableOperationContext))
     {
-        Log_Error("Failed handling agentinfo response.");
+        Log_Error("Fail handling agentinfo response: correlationId '%s'", messageContext->correlationId);
         goto done;
     }
 
 done:
+    free(correlationData);
     return;
 }
 
+void OnPublish_ainfo_resp(struct mosquitto* mosq, void* obj, const mosquitto_property* props, int reason_code)
+{
+    UNREFERENCED_PARAMETER(mosq);
+    UNREFERENCED_PARAMETER(props);
+
+    ADUC_Retriable_Operation_Context* retriable_operation_context = RetriableOperationContextFromAgentInfoMqttLibCallbackUserObj(obj);
+
+    switch (reason_code)
+    {
+    case MQTT_RC_NO_MATCHING_SUBSCRIBERS:
+        // No Subscribers were subscribed to the topic we tried to publish to (as per mqtt 5 spec).
+        // This is unexpected since at least the ADU service should be subscribed to receive the
+        // agent topic's publish. Set timer and try again later in hopes that the service will be
+        // subscribed, but fail and restart after max retries.
+        if (retriable_operation_context != NULL)
+        {
+            retriable_operation_context->retryFunc(retriable_operation_context, retriable_operation_context->retryParams);
+        }
+        break;
+
+    case MQTT_RC_UNSPECIFIED: // fall-through
+    case MQTT_RC_IMPLEMENTATION_SPECIFIC: // fall-through
+    case MQTT_RC_NOT_AUTHORIZED:
+        // Not authorized at the moment but maybe it can auto-recover with retry if it is corrected.
+        if (retriable_operation_context != NULL)
+        {
+            retriable_operation_context->retryFunc(retriable_operation_context, retriable_operation_context->retryParams);
+        }
+        break;
+
+    case MQTT_RC_TOPIC_NAME_INVALID: // fall-through
+    case MQTT_RC_PACKET_ID_IN_USE: // fall-through
+    case MQTT_RC_PACKET_TOO_LARGE: // fall-through
+    case MQTT_RC_QUOTA_EXCEEDED:
+        if (retriable_operation_context != NULL)
+        {
```

Comment:
trace worthy?

Reply:
yes. Added with this commit: [add tracing for puback failure and clang-format](https://github.com/Azure/iot-hub-device-update/pull/584/commits/ba38bbca78bf548ccec83aa3ef9cd0cc93b6c9e1)